packages/arb-ts/src/lib/utils/arbProvider.ts

```diff
@@ -0,0 +1,246 @@
+
+type ArbFormats = Formats & {
+  feeStats: FormatFuncs
+  feeStat: FormatFuncs
+  batchInfo: FormatFuncs
+}
+
+class ArbFormatter extends Formatter {
+  readonly formats!: ArbFormats
+
+  public getDefaultFormats(): ArbFormats {
+    // formats was already initialised in super, so we can just access here
+    const superFormats = super.getDefaultFormats()
+
+    const address = this.address.bind(this)
+    const bigNumber = this.bigNumber.bind(this)
+    const data = this.data.bind(this)
+    const hash = this.hash.bind(this)
+    const number = this.number.bind(this)
+    const feeStats = this.feeStats.bind(this)
+    const feeStat = this.feeStat.bind(this)
+    const batchInfo = this.batchInfo.bind(this)
+    const returnCode = this.returnCode.bind(this)
+
+    const arbTransactionFormat = {
+      ...superFormats.transaction,
+
+      l1SequenceNumber: bigNumber,
+      parentRequestId: hash,
+      indexInParent: number,
+      arbType: number,
+      arbSubType: number,
+      l1BlockNumber: number,
+    }
+
+    const arbReceiptFormat = {
+      ...superFormats.receipt,
+      returnData: data,
+      returnCode: returnCode,
+      feeStats: feeStats,
+      batchInfo: Formatter.allowNull(batchInfo, null),
+      l1BlockNumber: number,
+    }
+
+    const feeStatsFormat = {
+      prices: feeStat,
+      unitsUsed: feeStat,
+      paid: feeStat,
+    }
+
+    const feeStatFormat = {
+      l1Transaction: bigNumber,
+      l1Calldata: bigNumber,
+      l2Storage: bigNumber,
+      l2Computation: bigNumber,
+    }
+
+    const batchInfoFormat = {
+      confirmations: number,
+      blockNumber: number,
+      logAddress: address,
+      logTopics: Formatter.arrayOf(hash),
+      logData: data,
+    }
+
+    return {
+      ...superFormats,
+      transaction: arbTransactionFormat,
+      receipt: arbReceiptFormat,
+      feeStats: feeStatsFormat,
+      feeStat: feeStatFormat,
+      batchInfo: batchInfoFormat,
+    }
+  }
+
+  public returnCode(value: any): ReturnCode {
+    const bn = BigNumber.from(value)
+    const returnNum = bn.toNumber()
+    if (!Object.values(ReturnCode).includes(returnNum)) {
+      return logger.throwArgumentError('invalid return code', 'value', value)
+    }
+    return returnNum
+  }
+
+  public feeStat(feeStat: any): FeeStat {
+    return Formatter.check(this.formats.feeStat, feeStat)
+  }
+
+  public feeStats(feeStats: any): FeeStats {
+    return Formatter.check(this.formats.feeStats, feeStats)
+  }
+
+  public batchInfo(batchInfo: any): BatchInfo {
+    return Formatter.check(this.formats.batchInfo, batchInfo)
+  }
+
+  public transactionResponse(transaction: any): ArbTransactionResponse {
+    return super.transactionResponse(transaction) as ArbTransactionResponse
+  }
+
+  public receipt(value: any): ArbTransactionReceipt {
+    return super.receipt(value) as ArbTransactionReceipt
+  }
+}
+
+const getBatch = async (
+  seqNum: BigNumber,
+  l1BlockNumber: number,
```

Comment:
Can we rename this param so it's clear that the caller of getBatch is responsible for providing an l1BlockNumber that corresponds to the receipt the sequencer gave for the tx with the provided seqnum?

(Maybe "l2TxnSeqNum" and "l2TxL1BlockNumber"

Reply:
renamed, and added comment